django_mongodb_backend/cache.py

```diff
@@ -0,0 +1,203 @@
+    def _db(self):
+        return connections[router.db_for_read(self.cache_model_class)]
+
+    @property
+    def collection(self):
+        return self._db.get_collection(self._collection_name)
+
+    def get(self, key, default=None, version=None):
+        result = self.get_many([key], version)
+        if result:
+            return result[key]
+        return default
+
+    def _filter_expired(self, expired=False):
+        not_expired_filter = [{"expires_at": {"$gte": datetime.utcnow()}}, {"expires_at": None}]
+        operator = "$nor" if expired else "$or"
+        return {operator: not_expired_filter}
+
+    def get_many(self, keys, version=None):
+        if not keys:
+            return {}
+        keys_map = {self.make_and_validate_key(key, version=version): key for key in keys}
+        with self.collection.find(
+            {"key": {"$in": tuple(keys_map)}, **self._filter_expired(expired=False)}
+        ) as cursor:
+            return {keys_map[row["key"]]: self.serializer.loads(row["value"]) for row in cursor}
+
+    def set(self, key, value, timeout=DEFAULT_TIMEOUT, version=None):
+        key = self.make_and_validate_key(key, version=version)
+        serialized_data = self.serializer.dumps(value)
+        num = self.collection.count_documents({})
+        if num >= self._max_entries:
+            self._cull(num)
+        return self.collection.update_one(
+            {"key": key},
+            {
+                "$set": {
+                    "key": key,
+                    "value": serialized_data,
+                    "expires_at": self._get_expiration_time(timeout),
+                }
+            },
+            True,
+        )
+
+    def add(self, key, value, timeout=DEFAULT_TIMEOUT, version=None):
+        key = self.make_and_validate_key(key, version=version)
+        serialized_data = self.serializer.dumps(value)
+        num = self.collection.count_documents({})
+        if num >= self._max_entries:
+            self._cull(num)
+        try:
+            self.collection.update_one(
+                {"key": key, **self._filter_expired(expired=True)},
+                {
+                    "$set": {
+                        "key": key,
+                        "value": serialized_data,
+                        "expires_at": self._get_expiration_time(timeout),
+                    }
+                },
+                True,
+            )
+        except DuplicateKeyError:
+            # Check the exception name to catch when the key exists.
+            return False
+        return True
+
+    def _cull(self, num):
+        if self._cull_frequency == 0:
+            self.clear()
+        else:
+            keep_num = num - num // self._cull_frequency
+            try:
+                # Delete the first expiration date.
+                deleted_from = next(
+                    self.collection.aggregate(
+                        [
+                            {"$sort": SON([("expires_at", -1), ("key", 1)])},
```

Comment:
```suggestion
                            {"$sort": SON([("expires_at", -1), ("key", 1)])},
```

This can be a dictionary

Reply:
The son thing? I need to sort by expires_at and the by key name. will the dictionary hold the order?

Reply:
will change!